### ISSUE TITLE
No base count matrix(.tsv) files in Step 2 output but bam files of those cell types are there in Step1 Output.

### ISSUE THREAD

**Issue by lipika487** [cortes-ciriano-lab/SComatic#31]
I am trying to detect somatic mutations from my single-cell RNA-seq data for lung cancer samples. 
Step 1 command used: 
<img width="605" alt="image" src="https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/09f830bc-ba7e-4f46-aafa-3dcd8209f0aa">
Output files: 
![image](https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/f720a48a-9c9d-4a45-b58a-e4572c0c6e63)
Step 2:
<img width="548" alt="image" src="https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/8c9ce4e3-5f89-4dd8-a578-a85c51380786">

Output:
<img width="729" alt="image" src="https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/ce3717ee-038d-440c-a1b2-cde7c17d1c29">

I am not sure why there is no output files for all other cell types. Is there a specific reason that 10 out of 12 files are not there.
But all the files are processed (attaching the ss of subset of output of Step2)
<img width="864" alt="image" src="https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/357d862e-7b89-467a-9edc-4f88eec2eb83">

Note: I also tried using the Example data and its the same issue:
<img width="897" alt="image" src="https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/5411e5b5-2c14-43a1-886b-8c8ce140d075">
Only here, one file is missing.

**Comment by RinconFer** [comment]
Hi I am finding the same problem here, The step 1 runs with no problems and I have the bams files, however the step 2 code is generating the temporary folders (I have removed in the code the  rm -rf $temp line to check what was going on on those folders so I am keeping them, but they are empty).

However, it never adds anything inside the temp folders and the output file is not produced. It doesn't throw an error anywhere in the process either.
![image](https://github.com/cortes-ciriano-lab/SComatic/assets/62716429/d290084b-33f1-4ff4-8232-92214c7a051e)

**Comment by lipika487** [comment]
Yes, I tried it with other sample also, no file is produced now in Step 2 output even though it runs without any error. Also, all my temp files are empty.

**Comment by Francesc-Muyas** [comment]
Dear users,
Could you check couple of things? 

1. Could you show the size of the bam files. For instance using `ls -lh *bam`. 
2. Could you paste here the output (***report.txt**) of the _SplitBamCellTypes.py_ ?
3. Could you run the BaseCellCounter.py with the parameter `--nprocs 1` and print the error?
4. Finally, could you confirm that the **reference genome** version used in the SComatic computation is the same as the one used for the alignment (CellRanger)? 

Thanks, 
Fran

**Comment by lipika487** [comment]
1. 
<img width="589" alt="image" src="https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/3206aa90-00fc-4fdf-83e8-fd89a93200b8">

2. Sample report
<img width="831" alt="image" src="https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/cd40ea5b-ccc5-4c68-aa9f-ed89526f04b0">

3. I ran with nprocs=1 , and there is no error, it ran successfully and output of the same is attached in the issue also.
![image](https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/c4379afb-eeba-409e-9d69-35914cf761b3)
(I just printed the cell type extra here)

4. Yes, the reference genome version is same as used in the Cellranger for alignment.

**Comment by Francesc-Muyas** [comment]
Hi, 
It seems that the first step (_SplitBamCellTypes.py_) filtered a huge number of reads (66M out of 70M) because of CBs not being found in the metadata file. As consequence, step2 files are empty because there are almost no _PASS_ reads to analyse (see column Pass_reads in the report.txt file).  The CBs provided in the medatada file do rarely match with the ones observed in the bam file. 

Cheers,
Fran

**Comment by lipika487** [comment]
> Hi, It seems that the first step (_SplitBamCellTypes.py_) filtered a huge number of reads (66M out of 70M) because of CBs not being found in the metadata file. As consequence, step2 files are empty because there are almost no _PASS_ reads to analyse (see column Pass_reads in the report.txt file). The CBs provided in the medatada file do rarely match with the ones observed in the bam file.
> 
> Cheers, Fran

I see the issue now, thank you for your reply and help.
Is there any way I can see which CB are found, which are mismatched or not found?

Best,
Lipika

**Comment by Francesc-Muyas** [comment]
Hi, 

You could `grep` the CB columns in the bam file using this command `samtools view scRNAseq.bam | grep -o 'CB:Z:[ACTGN]*-1' | sort | uniq > unique.cb.txt` and then compare the output file with your metadata. 

You might need to slightly change the previous command depending on how your CBs look like. 

Thanks,
Fran

**Comment by lipika487** [comment]
> Hi, It seems that the first step (_SplitBamCellTypes.py_) filtered a huge number of reads (66M out of 70M) because of CBs not being found in the metadata file. As consequence, step2 files are empty because there are almost no _PASS_ reads to analyse (see column Pass_reads in the report.txt file). The CBs provided in the medatada file do rarely match with the ones observed in the bam file.
> 
> Cheers, Fran

Hi Fran, I tried with my other sample, in this case there is almost 85% of passed reads, still there is no file in Step 2. I am attaching the ss here. Also I could not see the unique CB of these bam files.
<img width="945" alt="image" src="https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/cae2ac13-2089-4650-a5e3-6b09ef589508">

**Comment by lipika487** [comment]
Hi Fran,
I used the same command for Endothelial cells.
I managed to get past this error by using meta file from annotating raw feature matrix without doing any QC or any filtering on it instead of doing the annotation later, so that no reads are removed.
<img width="859" alt="image" src="https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/03694c88-f7a0-448c-9c98-4303988497be">
I am guessing that the 4 cell types did not have enough passed reads, that's  why those 4 cell types are missing in Step 2. 
If I do Step 3 now, there is no file in this output even it completes without any error.
![image](https://github.com/cortes-ciriano-lab/SComatic/assets/84702870/5f2e96f1-a042-445c-9c46-d1797d0b9e53)

**Comment by Francesc-Muyas** [comment]
Could you paste the command you used to run the step 3? Why does it say that there are 15 tsv files in this folder? There should be only the ones you want to be merged for this sample/individual.

**Comment by lipika487** [comment]
Yes, thanks, it worked now. It was hidden.

> Have you check the hidden files in your folder? (Ctrl + H). It happened to me that I forgot to define the Sample object in the terminal and the script generated a file that started with a period, which in Linux it makes the file hidden



> Could you paste the command you used to run the step 3? Why does it say that there are 15 tsv files in this folder? There should be only the ones you want to be merged for this sample/individual.

Yes, i was guessing 15 files are the unique cell types (base count matrix) tsv files, that will be merged into Step3.
 
I successfully ran the SComatic without any errors.  Thankyou so much for all the help!

One more question, how to visualise these results graphically? (Sorry, I am very new to scRNA analysis :( )
Thanks in advance!

**Comment by Francesc-Muyas** [comment]
We used to perform the downstream analysis using R or Python depending on your priorities. However, this is not included in SComatic. 

Cheers,
Fran

**Comment by lipika487** [comment]
Thank you so much for your help :)
